crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -4090,8 +4090,13 @@ impl<'db> TypeInferenceBuilder<'db> {
                 // Class scopes are not visible to nested scopes, and we need to handle global
                 // scope differently (because an unbound name there falls back to builtins), so
                 // check only function-like scopes.
+                // However, there are exceptions to this rule: deferred types are visible,
+                // and type parameter scopes are evaluated lazily, thus class scopes can be visible.
                 let enclosing_scope_id = enclosing_scope_file_id.to_scope_id(db, current_file);
-                if !enclosing_scope_id.is_function_like(db) {
+                if !enclosing_scope_id.is_function_like(db)
+                    && !self.is_deferred()
+                    && !scope.is_annotation(db)
+                {
```

Comment:
As discussed in the other comment in the test suite, the exception you make here for `self.is_deferred()` I think we should not make.

And I think the `scope.is_annotation` exception needs to be a bit more precise. Annotation scopes can see names defined in their immediately enclosing class scope, but not names defined in more distant enclosing class scopes. For example:

```py
 class A:
    a = 1
    class B:
        b = 2
        type S = a  # error!
        type T = b  # no error
```

We should add a test showing that the `type S = a` case is an `unresolved-reference` error, and adjust the implementation here accordingly.